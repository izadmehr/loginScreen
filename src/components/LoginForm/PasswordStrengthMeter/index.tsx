import React from "react";
import { ZXCVBNScore } from "zxcvbn";

import PasswordStrengthMeterContianer from "./Styles";

interface Props {
  score: ZXCVBNScore;
  showMeterLabel: boolean;
  notSecure: boolean;
}
function createPasswordLabel(score: ZXCVBNScore): string {
  switch (score) {
    case 0:
      return "Weak";
    case 1:
      return "Weak";
    case 2:
      return "Fair";
    case 3:
      return "Good";
    case 4:
      return "Strong";
    default:
      return "Weak";
  }
}

function PasswordStrengthMeter(props: Props): JSX.Element {
  const { score, showMeterLabel } = props;
  const passwordLabel = createPasswordLabel(score);

  return (
    <PasswordStrengthMeterContianer>
      {showMeterLabel && (
        <progress
          className={`password-strength-meter-progress strength-${passwordLabel}`}
          value={score}
          max="4"
        />
      )}
      <br />
      <div className="password-strength-meter-label">
        {showMeterLabel && (
          <>
            <strong>Password strength:</strong> {passwordLabel}
          </>
        )}
      </div>
    </PasswordStrengthMeterContianer>
  );
}

export default PasswordStrengthMeter;
