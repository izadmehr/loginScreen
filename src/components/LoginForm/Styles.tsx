import styled, { css } from "styled-components";

import { colors } from "../../utlis/constants";

const btn = (light, dark) => css`
  height: 2.5em;
  white-space: nowrap;
  display: inline-block;
  border-radius: 2px;
  padding: 5px 1em;
  font-size: 1rem;
  color: white;
  &:visited {
    color: white;
  }
  background-color: ${light};
  border: 1px solid ${dark};
  &:hover {
    background-image: linear-gradient(${light}, ${dark});
    &[disabled] {
      background-image: linear-gradient(${light}, ${dark});
    }
  }
  &:visited {
    color: black;
  }
  &[disabled] {
    opacity: 0.6;
    cursor: not-allowed;
  }
`;

const btnPrimary = btn(colors.primary, colors.primaryDark);

export default styled.div`
  font-family: sans-serif;
  background-color: #f0f4f7;
  height: 100vh;
  form {
    width: 32rem;
    margin: auto;
    padding: 3rem 2rem;
    box-shadow: 0 1px 10px 0 rgba(0, 0, 0, 0.1);
    border-radius: 3px;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    background-color: #ffffff;

    .logo {
      width: 9rem;
      margin: 0 auto;
      display: block;
    }
    @media (max-width: 500px) {
      width: 100%;
      margin: auto;
      height: 100%;
    }

    & > div {
      line-height: 2em;
      margin: 5px;
      & > label {
        color: #333;
        font-size: 1rem;
        line-height: 1.5;
        & > input {
          display: block;
          width: 95%;
          padding: 0.375rem 0.75rem;
          font-size: 1rem;
          line-height: 1.5;
          color: #495057;
          background-color: #e8f0fe;
          background-clip: padding-box;
          border: 1px solid #ced4da;
          border-radius: 0.25rem;
          transition: border-color 0.15s ease-in-out,
            box-shadow 0.15s ease-in-out;
          :focus {
            border: 1px solid ${colors.primary};
            outline-width: 0;
            background-color: #ffffff;
          }
        }
      }
      & > .error {
        color: #d0011b;
        margin-top: 0.5em;
        border-radius: 3px;
        min-height: 1.25rem;
        line-height: 1.25rem;
      }

      & > div {
        & > label {
          display: block;
          & > input {
            margin-right: 3px;
          }
        }
      }
    }

    button {
      &[type="submit"] {
        ${btnPrimary};
      }
      width: 100%;
    }
  }
`;
