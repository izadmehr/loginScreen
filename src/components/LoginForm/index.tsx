import React from "react";
import { Form, Field } from "react-final-form";
import zxcvbn, { ZXCVBNScore } from "zxcvbn";

import Styles from "./Styles";
import PasswordStrengthMeter from "./PasswordStrengthMeter";

interface Values {
  email: string;
  password: string;
}

interface Errors {
  email?: string;
  password?: {
    score: ZXCVBNScore;
    message: string;
  };
}
export function validateEmail(email: string): boolean {
  // eslint-disable-next-line max-len,no-useless-escape
  const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

  return re.test(String(email).toLowerCase());
}

// eslint-disable-next-line @typescript-eslint/no-var-requires
const projectALogo = require("../../assets/img/projecta-logo.png");

let passwordScore: ZXCVBNScore;
const LoginForm = (): JSX.Element => (
  <Styles>
    <Form
      onSubmit={(values: Values) => alert(JSON.stringify(values, null, 2))}
      initialValues={{ email: "", password: "" }}
      validate={(values: Values): Errors => {
        const errors: Errors = {};

        if (!validateEmail(values.email)) {
          errors.email = "Wrong email address";
        }

        const passwordResult = zxcvbn(values.password || "");

        passwordScore = passwordResult.score;

        if (passwordResult.score <= 2) {
          errors.password = {
            score: passwordResult.score,
            message: passwordResult.feedback.warning
          };
        }

        return errors;
      }}
      render={({ handleSubmit, submitting }): JSX.Element => (
        <form onSubmit={handleSubmit}>
          <img className="logo" src={projectALogo} alt="Project A logo" />
          <Field name="email">
            {({ input, meta: { active, error, touched } }): JSX.Element => (
              <div>
                <label htmlFor="email">
                  Email:
                  <input
                    {...input}
                    type="text"
                    placeholder="example@test.com"
                    id="email"
                  />
                </label>
                <div className="error">{touched && !active ? error : null}</div>
              </div>
            )}
          </Field>

          <Field name="password">
            {({ input, meta: { active, error, touched } }): JSX.Element => (
              <div>
                <label htmlFor="password">
                  Password:
                  <input
                    {...input}
                    type="text"
                    placeholder="********"
                    id="password"
                  />
                </label>
                <PasswordStrengthMeter
                  score={passwordScore}
                  showMeterLabel={touched || active}
                  notSecure={!!error}
                />
                <div className="error">
                  {error && (touched || active) ? error.message : null}
                </div>
              </div>
            )}
          </Field>
          <button type="submit" disabled={submitting}>
            Submit
          </button>
        </form>
      )}
    />
  </Styles>
);

export default LoginForm;
