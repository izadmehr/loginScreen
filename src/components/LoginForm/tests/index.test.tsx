import React from "react";
import renderer from "react-test-renderer";

import LoginForm, { validateEmail } from "../index";

test("PasswordStrengthMeter render week password", () => {
  const tree = renderer.create(<LoginForm />).toJSON();

  expect(tree).toMatchSnapshot();
});

test("validate email: something@something.com", async () => {
  expect(validateEmail("something@something.com")).toBe(true);
});
test("validate email: a@b.b", async () => {
  expect(validateEmail("a@b.b")).toBe(false);
});
test("validate email: somebody@example", async () => {
  expect(validateEmail("somebody@example")).toBe(false);
});
