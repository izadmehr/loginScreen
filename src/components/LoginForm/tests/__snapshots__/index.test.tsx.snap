// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordStrengthMeter render strong password 1`] = `
<div
  className="sc-bdVaJa yMIDh"
>
  <form
    onSubmit={[Function]}
  >
    <img
      alt="Project A logo"
      className="logo"
      src="test-file-stub"
    />
    <div>
      <label
        htmlFor="email"
      >
        Email:
        <input
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="example@test.com"
          type="text"
          value=""
        />
      </label>
      <div
        className="error"
      />
    </div>
    <div>
      <label
        htmlFor="password"
      >
        Password:
        <input
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="********"
          type="text"
          value=""
        />
      </label>
      <div
        className="sc-bwzfXH dDvQJq"
      >
        <br />
        <div
          className="password-strength-meter-label"
        />
      </div>
      <div
        className="error"
      />
    </div>
    <button
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`PasswordStrengthMeter render week password 1`] = `
<div
  className="sc-bdVaJa yMIDh"
>
  <form
    onSubmit={[Function]}
  >
    <img
      alt="Project A logo"
      className="logo"
      src="test-file-stub"
    />
    <div>
      <label
        htmlFor="email"
      >
        Email:
        <input
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="example@test.com"
          type="text"
          value=""
        />
      </label>
      <div
        className="error"
      />
    </div>
    <div>
      <label
        htmlFor="password"
      >
        Password:
        <input
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="********"
          type="text"
          value=""
        />
      </label>
      <div
        className="sc-bwzfXH dDvQJq"
      >
        <br />
        <div
          className="password-strength-meter-label"
        />
      </div>
      <div
        className="error"
      />
    </div>
    <button
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
